### PR TITLE
HV-1020 Support for additional payload for HibernateConstraintViolation

### PR DESCRIPTION
**Pull request by hferentschik**
A pull request following up on #439. Switching the new method in `HibernateConstraintViolation` to:

    <C> C getDynamicPayload(Class<C> type);

Also using the term 'dynamicPayload'

### CODE REVIEW

**Comment by julien-may** [comment]
are you also goig to do the documentation part? otherwise i still like to do that now. (i am in holidays, so i finally have time now)

**Comment by hferentschik** [comment]
> are you also goig to do the documentation part? 

done - 32763598d8b5bf405af9fd9de291e01d52aceaaa. WDYT?

From my perspective we are good to go. @gunnarmorling !?

File: documentation/src/main/asciidoc/ch11.asciidoc [hunk 1]
```diff
@@ -825,7 +860,7 @@ validator factory still is referenced by application code.
 ====
 
 [[section-time-provider]]
-=== Time providers for `@Future` and `@Past`
+=== Time providers for @Future and @Past
```
Comment (gunnarmorling) [inline documentation/src/main/asciidoc/ch11.asciidoc]: I don't mind, but why did you remove it? Does it look bad with the mark-up in titles?

File: engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * @since 5.3
+ */
+public interface HibernateConstraintViolation<T> extends ConstraintViolation<T> {
+	/**
+	 * @param type The type of payload to retrieve
+	 *
+	 * @return an instance of the specified type set by the user via
+	 * {@link org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext#withDynamicPayload(Object)}.
+	 */
+	<C> C getDynamicPayload(Class<C> type);
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java]: So it doesn't work super-well with parameterized types (e.g. `List`), requiring an unchecked cast as in the test. The alternative would be to use a type literal a la CDI, but I don't think it's worth that or even better from a usability POV.

**Comment by gunnarmorling** [comment]
Rebased and applied; Thanks, @julien-may, @hferentschik, that's a nice one!

File: documentation/src/main/asciidoc/ch11.asciidoc (hunk 1 above)
Comment (hferentschik) [inline documentation/src/main/asciidoc/ch11.asciidoc]: > Does it look bad with the mark-up in titles?

+1 Default rendering sucks with back-tics in the title. I hope asciidoctor will give you soon more control over how reference format and numbering looks

File: engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java (hunk 2 above)
Comment (hferentschik) [inline engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java]: > The alternative would be to use a type literal a la CDI, but I don't think it's worth that or even better from a usability POV.

Yeah, not optimal with parameterized types, but as you say, I don't think it is worth getting into more complicated approaches.

**Comment by hferentschik** [comment]
> Rebased and applied

Awesome.

**Comment by julien-may** [comment]
> Rebased and applied

Yay!
